src/rgw/rgw_data_sync.cc

```diff
@@ -1227,26 +1257,12 @@ class RGWDataSyncShardCR : public RGWCoroutine {
           }
           for (log_iter = log_entries.begin(); log_iter != log_entries.end(); ++log_iter) {
             ldout(sync_env->cct, 20) << __func__ << ":" << __LINE__ << ": shard_id=" << shard_id << " log_entry: " << log_iter->log_id << ":" << log_iter->log_timestamp << ":" << log_iter->entry.key << dendl;
-            if (!marker_tracker->index_key_to_marker(log_iter->entry.key, log_iter->log_id)) {
+            auto handle = marker_tracker->start(log_iter->entry.key, log_iter->log_id,
+                                                0, log_iter->log_timestamp);
+            if (!handle) {
               ldout(sync_env->cct, 20) << __func__ << ": skipping sync of entry: " << log_iter->log_id << ":" << log_iter->entry.key << " sync already in progress for bucket shard" << dendl;
-              marker_tracker->try_update_high_marker(log_iter->log_id, 0, log_iter->log_timestamp);
-              continue;
-            }
-            if (!marker_tracker->start(log_iter->log_id, 0, log_iter->log_timestamp)) {
-              ldout(sync_env->cct, 0) << "ERROR: cannot start syncing " << log_iter->log_id << ". Duplicate entry?" << dendl;
             } else {
-              /*
-               * don't spawn the same key more than once. We can do that as long as we don't yield
-               */
-              if (spawned_keys.find(log_iter->entry.key) == spawned_keys.end()) {
-                spawned_keys.insert(log_iter->entry.key);
-                spawn(new RGWDataSyncSingleEntryCR(sync_env, log_iter->entry.key, log_iter->log_id, marker_tracker, error_repo, false), false);
-                if (retcode < 0) {
-                  stop_spawned_services();
-                  drain_all();
-                  return set_cr_error(retcode);
-                }
-              }
+              spawn(new RGWDataSyncSingleEntryCR(sync_env, log_iter->entry.key, marker_tracker, handle, error_repo, false), false);
```

Comment:
@cbodley yes, but the difference is that if we squash multiple entries without yielding, for these entries there's no need for retry since we know the spawned entries (that were started earlier) hadn't started yet, so we're not racing. Now that we yield for each spawned entry, we need to set retry for the following entries.